lib/ttfunk/table/cff/charset.rb

```diff
@@ -0,0 +1,230 @@
+          else
+            @count = self.class.strings_for_charset_id(offset_or_id).size
+          end
+        end
+
+        def each
+          return to_enum(__method__) unless block_given?
+          # +1 adjusts for the implicit .notdef glyph
+          (count + 1).times { |i| yield self[i] }
+        end
+
+        def [](glyph_id)
+          return FIRST_GLYPH_STRING if glyph_id == 0
+          find_string(sid_for(glyph_id))
+        end
+
+        def offset
+          # Numbers from 0..2 mean charset IDs instead of offsets. IDs are
+          # basically pre-defined sets of characters.
+          #
+          # In the case of an offset, add the CFF table's offset since the
+          # charset offset is relative to the start of the CFF table. Otherwise
+          # return nil (no offset).
+          if offset_or_id > 2
+            offset_or_id + top_dict.cff_offset
+          end
+        end
+
+        # mapping is new -> old glyph ids
+        def encode(mapping)
+          # no offset means no charset was specified (i.e. we're supposed to
+          # use a predefined charset) so there's nothing to encode
+          return '' unless offset
+
+          sids = mapping.keys.sort.map { |new_gid| sid_for(mapping[new_gid]) }
+          ranges = TTFunk::BinUtils.rangify(sids)
+          range_max = ranges.map(&:last).max
+
+          range_bytes = if range_max > 0
+                          (Math.log2(range_max) / 8).floor + 1
+                        else
+                          # for cases when there are no sequences at all
+                          Float::INFINITY
+                        end
+
+          # calculate whether storing the charset as a series of ranges is
+          # more efficient (i.e. takes up less space) vs storing it as an
+          # array of SID values
+          total_range_size = (2 * ranges.size) + (range_bytes * ranges.size)
+          total_array_size = sids.size * element_width(:array_format)
+
+          [].tap do |result|
+            if total_array_size <= total_range_size
+              result << [format_int(:array_format)].pack('C')
+              result << sids.pack('n*')
+            else
+              fmt = range_bytes == 1 ? :range_format_8 : :range_format_16
+              element_fmt = element_format(fmt)
+              result << [format_int(fmt)].pack('C')
+
+              ranges.each do |range|
+                sid, num_left = range
+                result << [sid, num_left].pack(element_fmt)
+              end
+            end
+          end.join
+        end
+
+        private
+
+        def sid_for(glyph_id)
+          return 0 if glyph_id == 0
+
+          # rather than validating the glyph as part of one of the predefined
+          # charsets, just pass it through
+          return glyph_id unless offset
+
+          case format_sym
+          when :array_format
+            # zero is always .notdef, so adjust with - 1
+            @entries[glyph_id - 1]
```

Comment:
Several points to bring up here:

1. I tried adding `attr_writer`s but they felt awkward to use in assignments because they look like local variables, which strikes me as the same sort of error-proneness as ivar typos. I think I would prefer to always use ivars when assigning and `attr_reader`s when accessing, and keep assignments localized to the constructor (and in this case the `parse` method, which can be thought of as a constructor of sorts for these subtables).
2. Even if we decided to avoid ivar usage completely during assignment, one of the ivars I'm using here is `format` which is also a method on `Kernel`. Sure, we could shadow it with an `attr_reader`/`attr_writer` but that feels... wrong. Maybe I should just rename it?

Reply:
One way you can make assignments more explicit is to use `self`:
```ruby
self.entries = []
```

While this doesn't fundamentally fix the issue it can help readers distinguish local vars from attributes.

I'll leave it up to you.